src/Octoshift/Extensions/CommandExtensions.cs

```diff
@@ -0,0 +1,37 @@
+using System;
+using System.CommandLine;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OctoshiftCLI.Extensions;
+
+public static class CommandExtensions
+{
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    public static RootCommand AddCommands(this RootCommand rootCommand, ServiceProvider serviceProvider)
+    {
+        if (rootCommand is null)
+        {
+            throw new ArgumentNullException(nameof(rootCommand));
+        }
+
+        foreach (var commandType in Assembly.GetCallingAssembly().GetAllDescendantsOfCommandBase())
+        {
+            var command = commandType.CreateInstance<Command>();
+
+            command.SetHandler(async ctx =>
+            {
+                var commandArgsType = commandType.BaseType.GetGenericArguments()[0];
+                var commandArgs = ctx.BindArgs(command, commandArgsType);
+                var handler = commandType.GetMethod("BuildHandler").Invoke(command, new[] { commandArgs, serviceProvider });
```

Comment:
Wouldn't it be better to check if `handler` implements `ICommandHander` and throw a meaningful exception? It is going to throw a `NullReferenceException` if `GetMethod` returns `null` in case it cannot find `BuildHandler` method anyways.